### ISSUE TITLE
Lack of error checking in bam_sort.c means running out of disk space does not produce an error return

### ISSUE THREAD

**Issue by smowton** [samtools/samtools#470]
The return value from sam_write1 and sam_close in bam_merge_core2 are not checked, meaning an ENOSPC error is ignored and the sort process reports success. This means broken BAM files are not detected and failures downstream are harder to trace back to the actual source of the problem. For example, if I create a small partition /mnt/small I can run:

```
$ samtools sort ~/large.bam /mnt/small/sorted.bam
# Note no error reported
$ echo $?
0 # Success
$ samtools view sorted.bam.bam | wc -l
[E::sam_parse1] SEQ and QUAL are of different length
[W::sam_read1] parse error at line 225844
[main_samview] truncated file.
  225839
```

**Comment by daviesrob** [comment]
Pull request samtools#467 fixes this:

```
$ df -h /mnt
Filesystem      Size  Used Avail Use% Mounted on
nodev            52K     0   52K   0% /mnt
$ ./samtools sort -o /mnt/srt.bam -T /tmp/srt /tmp/mpileup.1.sam
[bam_sort_core] error writing final output.
$ echo $?
1
```
It also catches the case where there isn't enough room for temporary files:
```
$ ./samtools sort -o /mnt/srt.out -T /mnt/srt -m 100K /tmp/mpileup.1.sam
[bam_sort_core] error sorting blocks.
$ echo $?
1
```

Hopefully it will be merged in time for release 1.3.

**Comment by smowton** [comment]
Great, thanks!

**Comment by jmarshall** [comment]
PR #467 is now merged, and the error messages improved in 5f53e3cd184a47b51f921a18402fc8bad8a39877.
